espnet/nets/pytorch_backend/e2e_tts_fastspeech.py

```diff
@@ -0,0 +1,623 @@
+
+        if args.transformer_init == "pytorch":
+            return
+        # weight init
+        for p in self.parameters():
+            if p.dim() > 1:
+                if args.transformer_init == "xavier_uniform":
+                    torch.nn.init.xavier_uniform_(p.data)
+                elif args.transformer_init == "xavier_normal":
+                    torch.nn.init.xavier_normal_(p.data)
+                elif args.transformer_init == "kaiming_uniform":
+                    torch.nn.init.kaiming_uniform_(p.data, nonlinearity="relu")
+                elif args.transformer_init == "kaiming_normal":
+                    torch.nn.init.kaiming_normal_(p.data, nonlinearity="relu")
+                else:
+                    raise ValueError("Unknown initialization: " + args.transformer_init)
+        # bias init
+        for p in self.parameters():
+            if p.dim() == 1:
+                p.data.zero_()
+        # reset some modules with default init
+        for m in self.modules():
+            if isinstance(m, (torch.nn.Embedding, LayerNorm)):
+                m.reset_parameters()
+
+    @property
+    def attention_plot_class(self):
+        return TTSPlot
+
+    @property
+    def base_plot_keys(self):
+        """base key names to plot during training. keys should match what `chainer.reporter` reports
+
+        if you add the key `loss`, the reporter will report `main/loss` and `validation/main/loss` values.
+        also `loss.png` will be created as a figure visulizing `main/loss` and `validation/main/loss` values.
+
+        :rtype list[str] plot_keys: base keys to plot during training
+        """
+        plot_keys = ["loss", "l1_loss", "duration_loss"]
+        if self.use_scaled_pos_enc:
+            plot_keys += ["encoder_alpha", "decoder_alpha"]
+
+        return plot_keys
+
+
+class DurationCalculator(torch.nn.Module):
+    """Duration calculator using teacher model
+
+    :param e2e_tts_transformer.Transformer teacher_model: teacher auto-regressive Transformer
+    """
+
+    def __init__(self, teacher_model):
+        super(DurationCalculator, self).__init__()
+        if not isinstance(teacher_model, Transformer):
+            raise ValueError("teacher model should be the instance of e2e_tts_transformer.Transformer")
+        self.teacher_model = teacher_model
+
+    def forward(self, xs, ilens, ys, olens):
+        """Calculate duration of each inputs
+
+        :param torch.Tensor xs: batch of padded character ids (B, Tmax)
+        :param torch.Tensor ilens: list of lengths of each input batch (B)
+        :param torch.Tensor ys: batch of padded target features (B, Lmax, odim)
+        :param torch.Tensor ilens: list of lengths of each output batch (B)
+        :return torch.Tensor: batch of durations (B, Tmax)
+        """
+        att_ws = self._calculate_attentions(xs, ilens, ys, olens)
+        # TODO(kan-bayashi): fix this issue
+        # this does not work in multi-gpu case. registered buffer is not saved.
```

Comment:
you can use state-less style like `LSTMCell`.
```python
def forward(self, xs, ilens, ys, olens, diag_head_idx):
  ...
  return pad_list(...), diag_head_idx
```

Reply:
Wow Clever